src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+ *  particular file as subject to the "Classpath" exception as provided
+ *  by Oracle in the LICENSE file that accompanied this code.
+ *
+ *  This code is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.util.ArraysSupport;
+
+import java.lang.foreign.AddressLayout;
+import java.lang.foreign.GroupLayout;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.PaddingLayout;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.ValueLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Array;
+import java.lang.reflect.RecordComponent;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A record mapper that is matching a GroupLayout to match the components of a record.
+ *
+ * @param <T> the Record type
+ */
+public final class LayoutRecordMapper<T extends Record>
+        implements Function<MemorySegment, T> {
+
+    enum Allow {EXACT, BOXING, BOXING_NARROWING_AND_WIDENING}
+
+    public static final Allow ALLOW = Allow.EXACT;
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+    private static final MethodHandles.Lookup PUBLIC_LOOKUP = MethodHandles.publicLookup();
+
+    private final Class<T> type;
+    private final GroupLayout layout;
+    private final MethodHandle ctor;
+
+    public LayoutRecordMapper(Class<T> type,
+                              GroupLayout layout) {
+        this(type, layout, 0);
+    }
+
+    private LayoutRecordMapper(Class<T> type,
+                               GroupLayout layout,
+                               long offset) {
+        this.type = type;
+        this.layout = layout;
+
+        var recordComponents = Arrays.asList(type.getRecordComponents());
+        var recordComponentNames = recordComponents.stream()
```

Comment:
I have a feeling that the use of streams here is hiding a much simpler loop where we loop over the record components, we check that each of them have a corresponding layout in the Group. For instance:

```
Map<String, MemoryLayout> nameToLayoutMap = ...
for (RecordComponent rc : type.getRecordComponents()) {
    try {
         MemoryLayout memberLayout = layout.select(PathElement.groupElement(rc.name());
         nameToLayoutMap = nameToLayoutMap(memberLayout.name().get(), memberLayout);
    } catch (IllegalArgumentException ex) {
         throw new IllegalArgumentException("There is no mapping...");
    }
}
```